Add component stack to invalid element type warning

n3tr opened the pull request:
To resolve #7856 and #7859 seems no longer active

Add stack addendum to  Invalid element type warning

- It will get stack from constructed element if the element has `_source` (file and line number), the stack will be look like `In Unknown (at file.js:12)`.
- In case jsx babel transform isn't being used, we won't have location info (in `element._source`). we will use parent's generated stack instead.

**Sample Code**
```js
let Foo = {};
class Button extends React.Component {
  render() {
    return (
      <div>
        <Foo /> // Foo is invalid element
      </div>
    );
  }
}

class DangerButton extends React.Component {
  render() {
    return <Button />;
  }
}

var ExampleApplication = React.createClass({
  render: function() {
    return (
      <div>
        <DangerButton />
      </div>
    )
  }
});

ReactDOM.render(
  <ExampleApplication />,
  document.getElementById('container')
);
```

**Output (with babel transform)**
```
react.js:3683 Warning: React.createElement: type should not be null, undefined, boolean, or number. It should be a string (for DOM elements) or a ReactClass (for composite components). Check the render method of `Button`.
    in Unknown (at example.js:7)
    in Button (at example.js:15)
    in DangerButton (at example.js:23)
    in div (at example.js:22)
    in ExampleApplication (at example.js:30)
```

**Output (without _source from babel)**
```
react.js:3683 Warning: React.createElement: type should not be null, undefined, boolean, or number. It should be a string (for DOM elements) or a ReactClass (for composite components). Check the render method of `Button`.
    in Button (created by DangerButton)
    in DangerButton (created by ExampleApplication)
    in div (created by ExampleApplication)
    in ExampleApplication
```

We drop the first line which is `in Unknown (created by Button)` since it's not very useful.

spicyj commented on src/isomorphic/classic/element/__tests__/ReactElementValidator-test.js:
```diff
@@ -535,12 +536,21 @@ describe('ReactElementValidator', () => {
       'ReactClass (for composite components).'
     );
+    // The format is '    in Unknown (at ReactElementValidator-test.js:539)'
+    expectDev(lines[1]).toMatch(
```
Can you use normalizeCodeLocInfo like the other tests?

n3tr commented:
Wouldn't it be better to show "Check the render method of … (at file.js:12)" if we know the owner and show "Check your code at file.js:12." if we don't. It looks more useful to me in case I have many components in single file. What do you think?

spicyj commented:
I think it's not necessary because the name will already be in the stack below, and in the case of "render callbacks" the current owner will often not be the "correct" component that is responsible.

n3tr commented:
Rebased and updated the code. 

**Output:**

```
Warning: React.createElement: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: boolean. Check your code (at Button.js:8).
    in Button (at DangerButton.js:6)
    in DangerButton (at App.js:10)
    in div (at App.js:9)
    in App (at index.js:7)
```

gaearon commented:
I think it's easy to miss the actual place where error happens and go to the parent:

<img width="555" alt="screen shot 2017-01-09 at 15 19 37" src="https://cloud.githubusercontent.com/assets/810438/21771384/0e83059e-d67f-11e6-8864-295426eb02c0.png">

My eyes stop at `index.js:7` here. I didn't even notice `Check your code at` here.

gaearon commented:
Proposal: break it out with a double newline.

<img width="561" alt="screen shot 2017-01-09 at 15 23 26" src="https://cloud.githubusercontent.com/assets/810438/21771524/9b3aeccc-d67f-11e6-8f5d-dfe0dd195d2f.png">

We could do this for all warnings since people commonly miss this tip at the very end.

gaearon commented:
I’ll get this in for now and look at delimiting with newlines later for all warnings together.
Thank you so much!
